File: solvings/medium/add-two-numbers.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution
{
    public ListNode AddTwoNumbers(ListNode l1, ListNode l2)
    {
        var l1Items = GetListNodeValues(l1);
        var l2Items = GetListNodeValues(l2);

        var maxItemsLength = Math.Max(l1Items.Length, l2Items.Length);
        var n1Values = ExtendCollection(l1Items, maxItemsLength);
        var n2values = ExtendCollection(l2Items, maxItemsLength);

        var result = new List<int>();
        var remainder = 0;
        for (var i = 0; i < n1Values.Length; i++)
        {
            var (val, rem) = Add(n1Values[i] + remainder, n2values[i]);
            remainder = rem;
            result.Add(val);
        }
        if (remainder != 0)
            result.Add(remainder);

        var resultListNode = new ListNode(result[0]);
        if (result.Count > 1)
        {

            var curr = new ListNode(result[1]);
            resultListNode.next = curr;

            for (var i = 2; i < result.Count; i++)
            {
                curr.next = new ListNode(result[i]);
                curr = curr.next;
            }
        }

        return resultListNode;
    }

    int[] GetListNodeValues(ListNode l1)
    {
        var values = new List<int>();

        var currentListNodePtr = l1;
        while (true)
        {
            values.Add(currentListNodePtr.val);

            if (currentListNodePtr.next == null)
                break;

            currentListNodePtr = currentListNodePtr.next;
        }

        return values.ToArray();
    }

    (int val, int remainder) Add(int a, int b)
    {
        var sum = a + b;

        var remainder = sum / 10;
        var val = sum % 10;

        return (val, remainder);
    }

    int[] ExtendCollection(int[] values, int length)
    {
        var res = new int[length];
        values.CopyTo(res, 0);
        return res;
    }
}

File: solvings/medium/add-two-numbers.go

package solvings

type ListNode struct {
	Val  int
	Next *ListNode
}

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	l1Curr := l1
	l2Curr := l2
	carryover := 0

	addRes, remainder := addNodes(l1Curr, l2Curr, carryover)
	carryover = remainder
	l1Curr = l1Curr.Next
	l2Curr = l2Curr.Next

	result := addRes
	resultLastNode := result
	for l1Curr != nil || l2Curr != nil {

		addRes, remainder = addNodes(l1Curr, l2Curr, carryover)
		carryover = remainder

		resultLastNode.Next = addRes
		resultLastNode = resultLastNode.Next

		if l1Curr != nil {
			l1Curr = l1Curr.Next
		}
		if l2Curr != nil {
			l2Curr = l2Curr.Next
		}
	}

	if carryover != 0 {
		transfNode := new(ListNode)
		transfNode.Val = carryover
		resultLastNode.Next = transfNode
	}

	return result
}

func bitwisAddition(a, b int) (sum, carryover int) {
	total := a + b

	if total >= 10 {
		sum = total - 10
		carryover = 1
	} else {
		sum = total
		carryover = 0
	}

	return
}

func addNodes(a, b *ListNode, carryover int) (res *ListNode, remainder int) {
	aVal := 0
	if a != nil {
		aVal = a.Val
	}
	bVal := 0
	if b != nil {
		bVal = b.Val
	}

	s, remainder := bitwisAddition(aVal+carryover, bVal)

	res = new(ListNode)
	res.Val = s

	return
}
